sdk/core/azure-core/src/main/java/com/azure/core/http/rest/SyncRestProxy.java

```diff
@@ -0,0 +1,692 @@
+                + " Response body: " + bodyRepresentation;
+
+            result = new IOException(message, e);
+        }
+        return result;
+    }
+
+    /**
+     * Create a publisher that (1) emits error if the provided response {@code decodedResponse} has 'disallowed status
+     * code' OR (2) emits provided response if it's status code ia allowed.
+     *
+     * 'disallowed status code' is one of the status code defined in the provided SwaggerMethodParser or is in the int[]
+     * of additional allowed status codes.
+     *
+     * @param decodedResponse The HttpResponse to check.
+     * @param methodParser The method parser that contains information about the service interface method that initiated
+     * the HTTP request.
+     * @return An async-version of the provided decodedResponse.
+     */
+    private HttpDecodedResponse ensureExpectedStatus(final HttpDecodedResponse decodedResponse,
+        final SwaggerMethodParser methodParser, RequestOptions options) {
+        final int responseStatusCode = decodedResponse.getSourceResponse().getStatusCode();
+
+        // If the response was success or configured to not return an error status when the request fails, return the
+        // decoded response.
+        if (methodParser.isExpectedResponseStatusCode(responseStatusCode)
+            || (options != null && options.getErrorOptions().contains(ErrorOptions.NO_THROW))) {
+            return decodedResponse;
+        }
+
+        // Otherwise, the response wasn't successful and the error object needs to be parsed.
+        byte[] responseBytes = decodedResponse.getSourceResponse().getBodyAsByteArray().block();
+        if (responseBytes == null || responseBytes.length == 0) {
+            //  No body, create exception empty content string no exception body object.
+            throw new RuntimeException(instantiateUnexpectedException(
+                methodParser.getUnexpectedException(responseStatusCode), decodedResponse.getSourceResponse(),
+                null, null));
+        } else {
+            Object decodedBody =  decodedResponse.getDecodedBodySync(responseBytes);
+            // create exception with un-decodable content string and without exception body object.
+            throw new RuntimeException(instantiateUnexpectedException(
+                methodParser.getUnexpectedException(responseStatusCode),
+                decodedResponse.getSourceResponse(), responseBytes, decodedBody));
+        }
+    }
+
+    private Object handleRestResponseReturnType(final HttpDecodedResponse response,
+        final SwaggerMethodParser methodParser,
+        final Type entityType) {
+        if (TypeUtil.isTypeOrSubTypeOf(entityType, Response.class)) {
+            final Type bodyType = TypeUtil.getRestResponseBodyType(entityType);
+
+            if (TypeUtil.isTypeOrSubTypeOf(bodyType, Void.class)) {
+                response.getSourceResponse().getBody().ignoreElements().block();
```

Comment:
I think we should replace this with `response.getSourceResponse().close()` or `response.getSourceResponse().getContent().toBytes()`.
Or best would be to apply different strategy to make sure that response bytes are read and ignored depending on BinaryDataContent type.